pkg/scheduler/plugins/coescheduling/core/core.go

```diff
@@ -0,0 +1,618 @@
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return 0, PodGroupNotSpecified
+	}
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := pgMgr.cache.getGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return 0, PodGroupNotFound
+	}
+	// first add pod to the gang's WaitingPodsMap
+	gang.addAssumedPod(pod)
+
+	gangGroup := gang.getGangGroup()
+	allGangGroupSatisfied := true
+	// only the gang itself
+	if len(gangGroup) == 0 {
+		allGangGroupSatisfied = gang.isGangResourceSatisfied()
+	} else {
+		// check each gang group
+		for _, groupName := range gangGroup {
+			gangTmp := pgMgr.cache.getGangFromCache(groupName)
+			if gangTmp != nil {
+				if !gangTmp.isGangResourceSatisfied() {
+					allGangGroupSatisfied = false
+					break
+				} else {
+					// if we cannot find the gang's group-gang  from the gangCache,
+					// we assumed that this gang should wait until timeout,
+					// user should check if the group-gang exists or the spelling mistakes
+					allGangGroupSatisfied = false
+					break
+				}
+			}
+		}
+	}
+	if !allGangGroupSatisfied {
+		gang.setTimeoutStartTime(timeNowFn())
+		return gang.WaitTime, Wait
+	}
+	return 0, Success
+}
+
+// PostFilter
+// i. If strict-mode, we will set scheduleCycleValid to false and release all assumed pods.
+// ii. If non-strict mode, we will do nothing.
+func (pgMgr *PodGroupManager) PostFilter(ctx context.Context, pod *corev1.Pod, handle framework.Handle, pluginName string) (*framework.PostFilterResult, *framework.Status) {
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Unschedulable, "")
+	}
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := pgMgr.cache.getGangFromCache(gangId)
+	if gang == nil {
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Unschedulable, fmt.Sprintf("can't find gang, gangName: %v, podName: %v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name)))
+	}
+
+	if gang.isGangResourceSatisfied() {
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Unschedulable, "")
+	}
+
+	if gang.getGangMode() == extension.GangModeStrict {
+		handle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+			waitingGangId := GetNamespaceSplicingName(waitingPod.GetPod().Namespace,
+				GetGangNameByPod(waitingPod.GetPod()))
+			if waitingGangId == gangId {
+				klog.Errorf("postFilter rejects the pod, gangName: %v, podName: %v",
+					gangId, GetNamespaceSplicingName(waitingPod.GetPod().Namespace, waitingPod.GetPod().Name))
+				waitingPod.Reject(pluginName, "gang rejection in PostFilter")
+			}
+		})
+		gang.setScheduleCycleValid(false)
+		// since we rejected the pod which is waiting in Permit stage, we need to clear the gang's TimeoutStartTime field
+		gang.clearTimeoutStartTime()
```

Comment:
TimeoutStartTime is strange,  may be “countdown” is better

Reply:
got it